File: examples/g1/cpp/galbot_robot/src/audio_example.cpp
#include <chrono>
#include <iomanip>
#include <iostream>
#include <fstream>
#include <limits>
#include <sstream>
#include <string>
#include <unordered_set>
#include <vector>
#include <thread>

#include "galbot_robot.hpp"

using namespace galbot::sdk;

// Added command help table structure
struct CommandInfo {
  std::string cmd;
  std::string description;
  std::string parameters;
  std::string example;
};

// Command help table
const std::vector<CommandInfo> COMMAND_TABLE = {
    {"start_microphone_stream_input", "Start microphone streaming audio input feature", "No parameters", "start_microphone_stream_input"},
    {"stop_microphone_stream_input", "Stop the specified microphone streaming audio input", "No parameters", "stop_microphone_stream_input"},
    {"write_audio_stream_output", "PCM audiodata audio output", "Audio file name to play (16K 16bit Mono PCM audio file path)", "write_audio_stream_output"},
    {"stop_audio_stream_output", "Stop the specified audio output stream or all active audio output streams", "No parameters", "stop_audio_stream_output"},
    {"set_volume", "Set system global volume", "Volume setting parameter (0-100)", "set_volume"},
    {"get_volume", "Get current system global volume", "No parameters", "get_volume"},
    {"help", "Show help", "No parameters", "help"},
    {"q", "Exit program", "No parameters", "q"}};

// command
void printCommandHelp() {
  std::cout << "\n================ Command Help Table =================" << std::endl;
  std::cout << std::left << std::setw(30) << "Command" << std::setw(30) << "Description" << std::setw(30) << "Parameters"
            << std::setw(30) << "example" << std::endl;
  std::cout << std::string(110, '-') << std::endl;

  for (const auto& cmd : COMMAND_TABLE) {
    std::cout << std::left << std::setw(40) << cmd.cmd << std::setw(60) << cmd.description << std::setw(40)
              << cmd.parameters << std::setw(30) << cmd.example << std::endl;
  }
  std::cout << "=============================================\n" << std::endl;
}


void save_audio_data(const std::shared_ptr<AudioData>& audio_data) {
  if (audio_data == nullptr) {
    std::cout << "audio_data is nullptr" << std::endl;
    return;
  }

  // std::cout << "Mic audio timestamp_ns: " << audio_data->header.timestamp_ns << std::endl;
  // std::cout << "frame_id is " << audio_data->header.frame_id << std::endl;
  // std::cout << "type is " << audio_data->type << std::endl;
  // std::cout << "format is " << audio_data->format << std::endl;
  // std::cout << "data size is " << audio_data->data.size() << std::endl;

  if (audio_data->type == "denoise_chunk" && audio_data->format == "pcm" && audio_data->data.size() > 0)
  {
    // std::cout << "save audio data:" << std::endl;
    std::ofstream outfile("denoise_audio.pcm", std::ios::app);
    outfile.write(reinterpret_cast<char *>(audio_data->data.data()), audio_data->data.size());
    // std::cout << "Audio saved to result_audio.pcm" << std::endl;
  }
}


// Safely read numeric inputs of multiple types and handle input errors
template<typename T>
bool safe_read_number(T& value, const std::string& prompt = "") {
  if (!prompt.empty()) {
    std::cout << prompt;
  }

  std::string input;
  std::getline(std::cin, input);

  // Check input stream status
  if (std::cin.fail() && !std::cin.eof()) {
    std::cin.clear();
    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    std::cout << "Input error, please enter a valid number." << std::endl;
    return false;
  }

  // Trim leading and trailing spaces
  input.erase(0, input.find_first_not_of(" \t"));
  input.erase(input.find_last_not_of(" \t") + 1);

  if (input.empty()) {
    std::cout << "Input error, please enter a valid number." << std::endl;
    return false;
  }

  try {
    size_t pos;
    T result;

    if (std::is_integral<T>()) {
      result = static_cast<T>(std::stoll(input, &pos));
    } else if (std::is_floating_point<T>()) {
      result = static_cast<T>(std::stold(input, &pos));
    }

    // Check whether the entire string has been converted
    if (pos != input.length()) {
      std::cout << "Input error, please enter a valid number." << std::endl;
      return false;
    }

    value = result;
    return true;
  } catch (const std::exception&) {
    std::cout << "Input error, please enter a valid number." << std::endl;
    return false;
  }
}


// Safely read filename input and handle input errors
bool safe_read_file_name(std::string& file_name, const std::string& prompt = "") {
  if (!prompt.empty()) {
    std::cout << prompt;
  }

  std::getline(std::cin, file_name);

  // Check input stream status
  if (std::cin.fail() && !std::cin.eof()) {
    std::cin.clear();
    std::cin.ignore();
  }
  return true;
}

int main(int argc, char* argv[]) {
  std::cout << "start" << std::endl;
  // Get the base_instance instance
  auto& base_instance = GalbotRobot::get_instance(MachineType::G1);

  if (!base_instance.init()) {
    std::cout << "base_instance init fail" << std::endl;
    return -1;
  }

  printCommandHelp();

  std::string cmd, stream_id;
  while (base_instance.is_running()) {
    std::cout << "Enter interface name to test:";
    std::string input_line;
    std::getline(std::cin, input_line);

    // Extract the first word as the command (ignore subsequent arguments)
    std::istringstream iss(input_line);
    iss >> cmd;

    // If input is empty, continue the loop
    if (cmd.empty()) {
      continue;
    }
    if (cmd == "help") {
      printCommandHelp();
      continue;
    } else if (cmd == "start_microphone_stream_input") {
      stream_id = base_instance.start_microphone_stream_input([](const std::shared_ptr<AudioData> audio_data){
        save_audio_data(audio_data);
      });

      std::cout << "Start microphone stream input, stream_id: "<< stream_id << std::endl;
    } else if (cmd == "stop_microphone_stream_input") {
      base_instance.stop_microphone_stream_input(stream_id);
      std::cout << "Stop microphone stream input, stream_id : " << stream_id << std::endl;
    } else if (cmd == "write_audio_stream_output") {
      std::string play_file;
      safe_read_file_name(play_file, "Please enter audio file name to play (16K 16bit Mono PCM audio file path): ");

      const auto chunk_size = 2560; // Data block size of 2560 bytes
      std::vector<char> audio_chunk(chunk_size, 0);
      std::cout << "Write audio stream data..." << std::endl;
      std::ifstream infile(play_file, std::ios::binary);
      if (!infile) {
        std::cout << "Failed to open for reading: " << play_file << std::endl;
        continue;
      } else {
        stream_id = "spk_stream_output_test";
        while (infile) {
          infile.read(audio_chunk.data(), static_cast<std::streamsize>(chunk_size));
          std::streamsize bytes_read = infile.gcount();
          if (bytes_read <= 0) break;
          std::string out_chunk(audio_chunk.data(), bytes_read);
          base_instance.write_audio_stream_output(out_chunk, stream_id);
          std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }
      }
      std::cout << "Write audio stream output, stream id : " << stream_id << std::endl;
    } else if (cmd == "stop_audio_stream_output") {
      base_instance.stop_audio_stream_output(stream_id);
      std::cout << "Stop audio stream output, stream id: " << stream_id << std::endl;
    } else if (cmd == "set_volume") {
      float volume = 60;
      if (!safe_read_number(volume, "Please enter volume setting parameter (0-100): ")) {
        continue;
      }

      if (!base_instance.set_volume(volume)) {
        std::cout << "Failed to set volume: " << volume << std::endl;
        continue;
      }
      std::cout << "Set current volume: " << volume << std::endl;
    } else if (cmd == "get_volume") {
      auto volume = base_instance.get_volume();
      std::cout << "Get current volume: " << volume << std::endl;
    } else if (cmd == "q") {
      break;
    } else {
      std::cout << "Invalid input" << std::endl;
    }
  }

  base_instance.request_shutdown();
  base_instance.wait_for_shutdown();
  base_instance.destroy();

  std::cout << "proccess end" << std::endl;
  return 0;
}

File: examples/g1/python/galbot_robot/audio_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import time
import struct
import math
import os
from galbot_sdk.g1 import GalbotRobot

# global variable
audio_callback_count = 0
last_audio_type = None

# Command help table
COMMAND_TABLE = [
    {
        "cmd": "start_microphone_stream_input",
        "description": "Start microphone streaming audio input feature",
        "parameters": "No parameters",
        "example": "start_microphone_stream_input"
    },
    {
        "cmd": "stop_microphone_stream_input",
        "description": "Stop the specified microphone streaming audio input",
        "parameters": "No parameters",
        "example": "stop_microphone_stream_input"
    },
    {
        "cmd": "write_audio_stream_output",
        "description": "PCM audiodata audio output",
        "parameters": "Enter audio file name to play (16K 16bit Mono PCM audio file path):",
        "example": "write_audio_stream_output"
    },
    {
        "cmd": "stop_audio_stream_output",
        "description": "Stop the specified audio output stream or all active audio output streams",
        "parameters": "No parameters",
        "example": "stop_audio_stream_output"
    },
    {
        "cmd": "set_volume",
        "description": "Set system global volume",
        "parameters": "Please enter volume setting parameter (0-100)",
        "example": "set_volume"
    },
    {
        "cmd": "get_volume",
        "description": "Get current system global volume",
        "parameters": "No parameters",
        "example": "get_volume"
    },
    {
        "cmd": "help",
        "description": "Show help",
        "parameters": "No parameters",
        "example": "help"
    },
    {
        "cmd": "q",
        "description": "Exit program",
        "parameters": "No parameters",
        "example": "q"
    }
]

# command
def print_command_help():
    print("\n================ Command Help Table =================")
    print(f"{'Command':<40} | {'Description':<30} | {'Parameters':<20} | {'example':<30}")
    print("-" * 130)
    for cmd in COMMAND_TABLE:
        print(
            f"{cmd['cmd']:<40} | {cmd['description']:<30} | {cmd['parameters']:<20} | {cmd['example']:<30}")
    print("=============================================\n")


# Safely read float input
def safe_read_float(prompt: str = "") -> float:
    """
    Safely get float input

    Args:
        prompt: prompt text

    Returns:
        Valid float input
    """
    if prompt:
        print(prompt, end='', flush=True)
    
    try:
        value = float(input())
        return value
    except ValueError:
        print("Input error, please enter a valid float.")
        return None
    except EOFError:
        print("\nInput ended")
        return None


# Safely read integer input
def safe_read_int(prompt: str = "") -> int:
    """
    Safely get integer input

    Args:
        prompt: prompt text

    Returns:
        Valid integer input
    """
    if prompt:
        print(prompt, end='', flush=True)
    
    try:
        value = int(input())
        return value
    except ValueError:
        print("Input error, please enter a valid integer.")
        return None
    except EOFError:
        print("\nInput ended")
        return None

# Safely read filename input and handle input errors
def safe_read_filename(prompt: str = "") -> str:
    """
    Safely get filename input

    Args:
        prompt: prompt text

    Returns:
        Valid filename input
    """
    if prompt:
        print(prompt, end='', flush=True)
    
    try:
        filename = input()
        return filename
    except ValueError:
        print("Input error, please enter a valid file name.")

# Save audio data
def save_audio_to_file(filename: str, audio_data: bytes):
    """
    Save audio data

    Args:
        filename: file name
        audio_data: audio data (bytes)
    """
    try:
        with open(filename, 'ab') as f:
            f.write(audio_data)
        # print(f"✓ Audio appended and saved to {filename}")
    except Exception as e:
        print(f"✗ Failed to save file: {e}")


# Read audio data from file
def read_audio_from_file(filename: str, chunk_size: int = 2560) -> list:
    """
    Read audio data from file

    Args:
        filename: file name
        chunk_size: chunk size (bytes)

    Returns:
        Audio data block list
    """
    try:
        audio_chunks = []
        with open(filename, 'rb') as f:
            while True:
                chunk = f.read(chunk_size)
                if not chunk:
                    break
                audio_chunks.append(chunk)
        print(f"✓ Read {len(audio_chunks)} chunks from {filename}")
        return audio_chunks
    except FileNotFoundError:
        print(f"✗ File not found: {filename}")
        return []
    except Exception as e:
        print(f"✗ Failed to read file: {e}")
        return []


# Audio callback function
def audio_callback(audio_data: dict):
    """
    Audio data callback function

    Args:
        audio_data: audio data dictionary including header, type, format, data
    """
    global audio_callback_count, last_audio_type
    
    if not audio_data:
        return

    header = audio_data.get("header", {})
    timestamp = header.get("timestamp", {})
    sec = timestamp.get("sec", 0)
    nanosec = timestamp.get("nanosec", 0)

    audio_type = audio_data.get("type", "unknown")
    audio_format = audio_data.get("format", "unknown")
    data_size = len(audio_data.get("data", b""))

    audio_callback_count += 1
    last_audio_type = audio_type

    # Print audio data info
    if audio_type == "vad_begin":
        print(f"\n[VAD_BEGIN] Voice activity detection started")
        pass
    elif audio_type == "vad_end":
        print(f"[VAD_END] Voice activity detection ended")
        pass
    elif audio_type == "vad_chunk":
        # print(f"[AUDIO_CHUNK] Received {data_size} bytes of {audio_format} audio data (timestamp: {sec}.{nanosec:09d})")
        save_audio_to_file("mic_vad.pcm", audio_data.get("data", b""))
        pass
    elif audio_type == "denoise_chunk":
        # print(f"[DENOISE_AUDIO] Received {data_size} bytes of denoised audio (timestamp: {sec}.{nanosec:09d})")
        save_audio_to_file("mic_denoise.pcm", audio_data.get("data", b""))
        pass
    elif audio_type == "waken_up":
        print(f"\n[WAKEN_UP] Wake-up event triggered: ", audio_data.get("data", ""))
        pass
    else:
        print(f"[{audio_type}] Received audio data, size: {data_size} bytes")
        pass


def main():
    # Get GalbotRobot singleton
    base_instance = GalbotRobot()

    ok = base_instance.init()
    if not ok:
        print("Initialization failed")
        return -1

    print("\n========================================")
    print("     Galbot audio test tool started")
    print("========================================\n")

    print_command_help()

    microphone_stream_id = ""
    speaker_stream_id = "speaker_audio_output_test"
    while base_instance.is_running():
        print("Enter operation to execute:", end='', flush=True)
        cmd_line = input().strip()

        if not cmd_line:
            continue

        parts = cmd_line.split()
        cmd = parts[0]

        try:
            if cmd == "help":
                print_command_help()

            elif cmd == "start_microphone_stream_input":
                print("Starting microphone streaming audio input...")
                global audio_callback_count, last_audio_type
                
                microphone_stream_id = base_instance.start_microphone_stream_input(
                    callback=audio_callback,
                    chunk_size=2560,
                    use_raw_audio=False
                )
                audio_callback_count = 0
                last_audio_type = None
                print(f"✓ Microphone started successfully, stream ID: {microphone_stream_id}")

            elif cmd == "stop_microphone_stream_input":
                if microphone_stream_id is None:
                    print("No active microphone stream")
                else:
                    print(f"Stopping microphone stream (ID: {microphone_stream_id})...")
                    base_instance.stop_microphone_stream_input(microphone_stream_id)
                    print(f"✓ Microphone stopped")
                    print(f"  - Number of callbacks: {audio_callback_count}")
                    print(f"  - Last audio type: {last_audio_type}")
                    microphone_stream_id = None

            elif cmd == "write_audio_stream_output":
                audio_output_file = safe_read_filename("Enter audio file name to play (16K 16bit Mono PCM audio file path): ")
                print("Reading audio file...")
                # Read and play audio
                audio_chunks = read_audio_from_file(audio_output_file, chunk_size=2560)
                if audio_chunks:
                    print(f"Playing audio stream ID: {speaker_stream_id}")
                    
                    for i, chunk in enumerate(audio_chunks):
                        success = base_instance.write_audio_stream_output(chunk, speaker_stream_id)
                        if not success:
                            print(f"✗ Playback failed for chunk {i+1}")
                            break

                        time.sleep(0.05)

                    print(f"✓ Audio playback completed")

            elif cmd == "stop_audio_stream_output":
                print("Stopping audio output stream...")
                base_instance.stop_audio_stream_output(speaker_stream_id)
                print("✓ Audio output stream stopped")

            elif cmd == "set_volume":
                volume = safe_read_float("Please enter volume setting parameter (0-100): ")
                if volume is None or volume < 0 or volume > 100:
                    print(f"✗ Volume must be between 0-100; input value: {volume}")
                    continue

                if (not base_instance.set_volume(volume)):
                    print(f"✗ Failed to set volume({volume})")
                print(f"✓ Volume set to: {volume}%")

            elif cmd == "get_volume":
                volume = base_instance.get_volume()
                print(f"Current system volume: {volume}%")

            elif cmd == "q":
                print("Program exiting...")
                break

            else:
                print(f"Unknown command: {cmd}, enter 'help' to see available commands")

        except (ValueError, TypeError) as e:
            print(f"Parameters error: {e}")
        except Exception as e:
            print(f"Command execution exception: {e}")

    base_instance.request_shutdown()
    base_instance.wait_for_shutdown()
    base_instance.destroy()
    print("Program exited")


if __name__ == "__main__":
    main()
